sql/schemachanger: automatically adjust the batch size of the merge

spilchen opened the pull request:
We have observed contention issues with the schema changer's merge process, usually manifesting as a WriteTooOldErrors in the kv. KV retries the transaction up to 100 times before failing. Although various settings can be adjusted to control this behaviour and allow the schema change to succeed, this approach requires manual intervention.

This update makes the schema changer aware of such contention and enables it to automatically adjust the batch size. Specifically, it will detect when the KV error hits the auto-retry limit, allowing the schema changer to respond to this error by reducing its batch size for the next merge attempt.

Previously, a roachtest (alterpk-bank) was modified to retry more often by tuning the `kv.transaction.internal.max_auto_retries` setting. This has been removed from the test now, as the new code automatically reduces the batch size, providing a reliable solution to the issue.

Closes: #126197
Release note (bug fix): Changed the schema changer’s merge process so that it can detect contention errors and automatically retry with a smaller batch size. This makes the merge process more likely to succeed without needing any manual tuning of settings.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128201)
<!-- Reviewable:end -->

rafiss reviewed (approved):
this is a great enhancement and an elegant refactoring. lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128201#-:-O3cazZs7NcUY98AQ3WO:bb83d8o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @spilchen, and @vidit-bhat)

___
*[`pkg/sql/backfill/mvcc_index_merger.go` line 576 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128201#-O3caL19BUHP75pSu6ON:-O3caL1A3Aa_NXiUQYwm:b-i6lyst) ([raw file](https://github.com/cockroachdb/cockroach/blob/627cef7eb43b76df5fe7866157a26b0a2dc8771c/pkg/sql/backfill/mvcc_index_merger.go#L576)):*
> ```Go
> 		// We stop retrying if we still couldn't complete the batch at the minimum
> 		// size. We will return the error and let a higher level handle it.
> 		if batch.batchSize == minBatchSize {
> ```

super nit: since `minBatchSize` is only used within this function, the `const minBatchSize = 1` declaration could be the first line of this function.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: automatically adjust the batch size of the merge

We have observed contention issues with the schema changer's merge
process, usually manifesting as a WriteTooOldErrors in the kv. KV
retries the transaction up to 100 times before failing. Although various
settings can be adjusted to control this behaviour and allow the schema
change to succeed, this approach requires manual intervention.

This update makes the schema changer aware of such contention and
enables it to automatically adjust the batch size. Specifically, it will
detect when the KV error hits the auto-retry limit, allowing the schema
changer to respond to this error by reducing its batch size for the next
merge attempt.

Previously, a roachtest (alterpk-bank) was modified to retry more often
by tuning the `kv.transaction.internal.max_auto_retries` setting. This
has been removed from the test now, as the new code automatically
reduces the batch size, providing a reliable solution to the issue.

Closes: #126197
Release note (bug fix): Changed the schema changer’s merge process so
that it can detect contention errors and automatically retry with a
smaller batch size. This makes the merge process more likely to succeed
without needing any manual tuning of settings.

spilchen reviewed (commented):
tftr

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128201#-:-O3cup9ICnwEY1PnUvfs:bb7p3w5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, and @vidit-bhat)

___
*[`pkg/sql/backfill/mvcc_index_merger.go` line 576 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128201#-O3caL19BUHP75pSu6ON:-O3cukGj42cifbnK51iz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/627cef7eb43b76df5fe7866157a26b0a2dc8771c/pkg/sql/backfill/mvcc_index_merger.go#L576)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: since `minBatchSize` is only used within this function, the `const minBatchSize = 1` declaration could be the first line of this function.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen commented:
tftr

bors r=rafiss
